src/test/java/org/commcare/formplayer/auth/HmacAuthTests.java

```diff
@@ -115,6 +118,14 @@ public void testFullAuthEndpoint_WithHmacAuthAndUserDetails_Succeeds() throws Ex
         this.testEndpoint(builder, status().isOk());
     }
 
+    @Test
+    public void testFullAuthMultipartEndpoint_WithHmacAuthAndUserDetails_Succeeds() throws Exception {
+        String hmac = RequestUtils.getHmac(formplayerAuthKey, FULL_AUTH_BODY);
```

Comment:
for multipart requests the HMAC should be calculated on the full body, not just the JSON part. I think you will have to re-factor the `RequestUtils` so that you can get both the full body and the JSON part:

```java
RequestUtils.getBody(request)  // returns the full body content and does't care about the content type
RequestUtils.getJsonBodyPart(request)  // returns the JSON request part which might be the full body or just a part of it
```